### ISSUE TITLE
LowCq register allocation improvements

### ISSUE THREAD

**Issue by gdkchan** [Ryujinx/Ryujinx#2688]
## Feature Request

The JIT currently has two register allocators, one is a linear scan register allocator (LSRA) and the other is called "hybrid allocator". LSRA is nothing new, and is not the topic of the discussion here. The hybrid allocator is supposed to be a *fast* allocator, faster than LSRA, and is used on the initial translation of the functions. The name comes from the fact that it was supposed to split the free reigster sets into two groups, one for local (inside a basic block) allocation, and another for global (across the entire function) allocation.

This is not what is happening right now, however. It currently functions as a local allocator, the variables that are only live inside a single basic block are assigned a register, that is then returned to the free register set after its last use. Variables that are live across multiple blocks are always spilled.

#### Overview:
My proposal is improving the hybrid allocator by also assigning registers to variables live across multiple blocks. There is a simple way to do that without decreasing the throughput too much.

First, we sort variables by uses count in decreasing order. Then, for the first N variables, we assign N registers from the free register set. The value of N would need to be adjusted based on testing, but I believe it should less than half the amount of free registers.

The idea here is that both the long lived variables and short ones are able to get a register. By doing that, we also take advantage of the register allocation done on the guest code, where it should be already using the lowest possible amount of registers.

Ultimately, this allows improving the code generation of LowCq code at a very little cost, so I believe it is worth it.

#### Smaller Details: 
It may be necessary to spill the registers for the long lived variables across calls. And it should probably avoid picking the "fixed" registers aswell.

#### Nature of Request:  
Code generation improvement (in LowCq mode).

### Why would this feature be useful?
It reduces the amount of spills in LowCq code, making it faster and more compact. It mostly benefits first time runs without PPTC, but other cases should benefit a bit aswell.

**Comment by FICTURE7** [comment]
I think it might be worth improving the local register allocation capabilities of the hybrid allocator first. Here is my understanding of the current state of things:

It will split the set of registers into a two sets: a spill set and a free set. The free set is used to store block locals which are not spilled. The spill set contains 4 registers and is used to "stage" spilled locals to and from the stack. When the local is not a block local or the free set is exhausted; it is spilled for its entire lifetime. The free set also does not contain caller saved registers when the block contains a call (presumably to simplify and speed up allocation).

While this approach is very fast and simplifies allocation a lot, having locals spilled for their entire lifetime can cause a fair bit of redundant spill. This spill code has to be generated in the assembler which decreases throughput in turn.

I think we should attempt to have a more precise tracking of the live range of locals, at least at the block level, I'll see if I can come up with something that does not destroy throughput.

**Comment by gdkchan** [comment]
> The free set also does not contain caller saved registers when the block contains a call (presumably to simplify and speed up allocation).

Yea, I think this is not a big deal right now because most block locals are short-lived, so it shouldn't use a large amount of registers at once. It is probably more problematic on linux since the ABI has a higher number of caller-saved registers there. Still, it should be simple enough to save/restore those registers across calls. And then, it can just prefer registers that are not caller-saved to avoid that being necessary.

> I think we should attempt to have a more precise tracking of the live range of locals, at least at the block level, I'll see if I can come up with something that does not destroy throughput.

How would it improve compared to the current solution? IIRC, currently it keep tracks of the uses count, and the live range starts at the first definition, and ends on the last use. This should be correct for a variable that is live inside a single block. The only improvement I can think here is tracking lifetime of variables spanning multiple blocks, is that what you meant?

**Comment by FICTURE7** [comment]
I mean in the block itself, for example: Suppose a spilled local is reloaded from the stack to a register (currently a spill temp), since the local is on the use side, we do not need to reload it again the next time it is in a use if the register is not modified.

It does something like this currently:

Before RA:
```llvm
  i64 %0 = Add i64 %1, i64 %2    ;; Suppose %1 is a spilled local
  i64 %3 = Add i64 %1, i64 %3
```

After RA:
```llvm
  i64 r0 = Fill i32 0x00
  i64 %0 = Add i64 r0, i64 %2
  i64 r0 = Fill i32 0x00 
  i64 %3 = Add i64 r0, i64 %3
```

In LCQ guest registers are not in SSA, this can happen quite a bit. This also causes us to not use the full free set because we have to go through the spill set.

We can track that `%1` is currently living in `r0` so the next time it is referred, we can just use that register instead. Then evict/spill `%1` in `r0` if it happens we hit a fixed register use or we've exhausted the free register set and need a register.

Here is a basic block generated in LCQ I have lying around, I do not have the IR dump though; would have provided some more insight on it.

```asm
.L4:
  mov rbp,[rsp+90h]
  mov rsi,PAGE_TABLE
  mov rbp,[rsi+rbp+7D0h]
  mov rax,rbp
  mov [rsp+54h],rax  ;; Weird spill/fill sequence. Presumably rax contains a guest register which is a non-block local.
  mov rax,[rsp+54h]  ;; Not sure why it did that. Maybe a `i64 %0 = Copy i64 %0` got folded followed by a spill (because of the def).
  mov [rsp+54h],rax
  mov rbp,[rsp+98h]  ;; This is not needed.
  mov rax,[rsp+54h]  ;; This is not needed.
  mov [rsi+rbp+8],rax
```

**Comment by gdkchan** [comment]
Right. So it would basically keep track of whenever a variable is on a register or stack, and move as needed, instead of the current "all or nothing" situation where it is either always on a register, or always on the stack (with the reserved temp register used for reload/store as there is no support for stack operands right now).
